sdks/unreal/Agones/Source/Agones/Private/AgonesComponent.cpp

```diff
@@ -0,0 +1,496 @@
+			if (!bSucceeded)
+			{
+				ErrorDelegate.ExecuteIfBound({"Unsuccessful Call"});
+				return;
+			}
+
+			SuccessDelegate.ExecuteIfBound({});
+		});
+	Request->ProcessRequest();
+}
+
+void UAgonesComponent::SetAnnotation(
+	FString& Key, FString& Value, const FSetAnnotationDelegate SuccessDelegate, const FAgonesErrorDelegate ErrorDelegate)
+{
+	const FKeyValuePair Label = {Key, Value};
+	FString Json;
+	if (!FJsonObjectConverter::UStructToJsonObjectString(Label, Json))
+	{
+		ErrorDelegate.ExecuteIfBound({FString::Format(TEXT("error serializing key-value pair ({0}: {1}})"), {*Key, *Value})});
+		return;
+	}
+
+	TSharedRef<IHttpRequest> Request = BuildAgonesRequest("metadata/annotation", FHttpVerb::Put, Json);
+	Request->OnProcessRequestComplete().BindLambda(
+		[SuccessDelegate, ErrorDelegate](FHttpRequestPtr HttpRequest, FHttpResponsePtr HttpResponse, const bool bSucceeded) {
+			if (!bSucceeded)
+			{
+				ErrorDelegate.ExecuteIfBound({"Unsuccessful Call"});
+				return;
+			}
+
+			SuccessDelegate.ExecuteIfBound({});
+		});
+	Request->ProcessRequest();
+}
+
+void UAgonesComponent::Allocate(const FAllocateDelegate SuccessDelegate, const FAgonesErrorDelegate ErrorDelegate)
+{
+	TSharedRef<IHttpRequest> Request = BuildAgonesRequest("allocate");
+	Request->OnProcessRequestComplete().BindLambda(
+		[SuccessDelegate, ErrorDelegate](FHttpRequestPtr HttpRequest, FHttpResponsePtr HttpResponse, const bool bSucceeded) {
+			if (!bSucceeded)
+			{
+				ErrorDelegate.ExecuteIfBound({"Unsuccessful Call"});
+				return;
+			}
+
+			SuccessDelegate.ExecuteIfBound({});
+		});
+	Request->ProcessRequest();
+}
+
+void UAgonesComponent::Reserve(
+	const int64 Seconds, const FReserveDelegate SuccessDelegate, const FAgonesErrorDelegate ErrorDelegate)
+{
+	const FDuration Duration = {Seconds};
+	FString Json;
+	if (!FJsonObjectConverter::UStructToJsonObjectString(Duration, Json))
+	{
+		ErrorDelegate.ExecuteIfBound({TEXT("Failed to serializing request")});
+		return;
+	}
+
+	TSharedRef<IHttpRequest> Request = BuildAgonesRequest("reserve", FHttpVerb::Post, Json);
+	Request->OnProcessRequestComplete().BindLambda(
+		[SuccessDelegate, ErrorDelegate](FHttpRequestPtr HttpRequest, FHttpResponsePtr HttpResponse, const bool bSucceeded) {
+			if (!bSucceeded)
+			{
+				ErrorDelegate.ExecuteIfBound({"Unsuccessful Call"});
+				return;
+			}
+
+			SuccessDelegate.ExecuteIfBound({});
+		});
+	Request->ProcessRequest();
+}
+
+void UAgonesComponent::PlayerConnect(
+	const FString PlayerId, const FPlayerConnectDelegate SuccessDelegate, const FAgonesErrorDelegate ErrorDelegate)
+{
+	const FAgonesPlayer Player = {PlayerId};
+	FString Json;
+	if (!FJsonObjectConverter::UStructToJsonObjectString(Player, Json))
+	{
+		ErrorDelegate.ExecuteIfBound({TEXT("Failed to serializing request")});
+		return;
+	}
+
+	// TODO(dom) - look at JSON encoding in UE4.
+	Json = Json.Replace(TEXT("playerId"), TEXT("playerID"));

```

Comment:
Any way this can be done differently?

Reply:
This frustrated me for days ... I would love to know if there is a way to fix this, seems that JSON encoding isnt great in UE4. Do you have any suggestions?

Reply:
Feel your pain.. Not sure either, does renaming the `PlayerId` field of `FAgonesPlayer` to `PlayerID` do anything?

Reply:
nope, I did try that ... could have just been holding it wrong but have had a few ppl look at it internally and nothing obvious came up. If you could see if the i8e team knows a work around that would be fantastic.